flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/logical/PushPartitionIntoTableSourceScanRule.scala

```diff
@@ -94,14 +104,47 @@ class PushPartitionIntoTableSourceScanRule extends RelOptRule(
       inputFieldType.getFieldList.get(index).getType
     }.map(FlinkTypeFactory.toLogicalType)
 
-    val allPartitions = tableSource.getPartitions
-    val remainingPartitions = PartitionPruner.prunePartitions(
-      call.getPlanner.getContext.asInstanceOf[FlinkContext].getTableConfig,
-      partitionFieldNames,
-      partitionFieldTypes,
-      allPartitions,
-      finalPartitionPredicate
-    )
+    def getAllPartitions: util.List[util.Map[String, String]] = {
+      catalog match {
+        case Some(c) =>
+          c.listPartitions(tableIdentifier.get.toObjectPath)
+            .map(_.getPartitionSpec).toList
+        case None => tableSource.getPartitions
+      }
+    }
+
+    def internalPartitionPrune(): util.List[util.Map[String, String]] = {
+      val allPartitions = getAllPartitions
+      PartitionPruner.prunePartitions(
+        config,
+        partitionFieldNames,
+        partitionFieldTypes,
+        allPartitions,
+        finalPartitionPredicate
+      )
+    }
+
+    val remainingPartitions: util.List[util.Map[String, String]] = catalog match {
+      case Some(catalog) =>
+        val converter = new RexNodeToExpressionConverter(
+          inputFields,
+          context.getFunctionCatalog,
+          context.getCatalogManager,
+          TimeZone.getTimeZone(config.getLocalTimeZone))
+        finalPartitionPredicate.accept(converter) match {
+          case Some(expression) =>
+            try {
+              catalog
+                  .listPartitionsByFilter(tableIdentifier.get.toObjectPath, Seq(expression))
+                  .map(_.getPartitionSpec)
+            } catch {
+              case _: UnsupportedOperationException =>
```

Comment:
If we rely on the exception to fall back to internal implementation, we'd better mention it in the API contract, so that implementing class knows what to do if it doesn't support the API.